### PR TITLE
Add: Import and export tool for csv/parquet/json

### PR DESCRIPTION
**Pull request by Arman-Ghazaryan**

### CODE REVIEW

File: tools/dataset.cpp [hunk 1]
```diff
@@ -1 +1,199 @@
...
+
+    auto sources = table->GetColumnByName(c.source_id_field);
+    return_if_error(sources, c.error, 0, fmt::format("{} is not exist", c.source_id_field).c_str());
+    auto targets = table->GetColumnByName(c.target_id_field);
+    return_if_error(targets, c.error, 0, fmt::format("{} is not exist", c.target_id_field).c_str());
+    auto edges = table->GetColumnByName(c.edge_id_field);
+    int count = sources->num_chunks();
+    array.reserve(std::min(ukv_size_t(sources->chunk(0)->length()), max_batch_size));
+
+    for (size_t chunk_idx = 0; chunk_idx != count; ++chunk_idx) {
```
Comment (ashvardanian) [inline tools/dataset.cpp:64]: signed vs unsigned equality check

File: tools/dataset.cpp [hunk 2]
```diff
@@ -1 +1,199 @@
...
+        }
+    }
+    if (array.size() != 0)
+        upsert_one_batch(c, array);
+}
+
+void import_parquet(ukv_graph_import_t& c, ukv_size_t max_batch_size) {
+
+    arrow::Status status;
+    arrow::MemoryPool* pool = arrow::default_memory_pool();
```
Comment (ashvardanian) [inline tools/dataset.cpp:91]: We can use our arenas, like in client

File: tools/dataset.cpp [hunk 3]
```diff
@@ -1 +1,199 @@
...
+
+    ukv_key_t source = 0;
+    ukv_key_t target = 0;
+    ukv_key_t edge = 0;
+
+    for (auto doc : docs) {
+        simdjson::ondemand::object data = doc.get_object().value();
+        source = rewinded(data)[c.source_id_field];
+        target = rewinded(data)[c.target_id_field];
+        edge = edge_state ? rewinded(data)[c.edge_id_field] : 0;
```
Comment (ashvardanian) [inline tools/dataset.cpp:173]: edge ids have default value, check constants in the header

File: tools/dataset.cpp [hunk 4]
```diff
@@ -1 +1,199 @@
...
+}
+
+simdjson::ondemand::object& rewinded(simdjson::ondemand::object& doc) noexcept {
+    doc.reset();
+    return doc;
+}
+
+void import_json(ukv_graph_import_t& c, ukv_size_t max_batch_size) {
+
+    std::vector<edge_t> array;
```
Comment (ashvardanian) [inline tools/dataset.cpp:148]: cant we preallocate a max size vector?

**Review by ashvardanian — CHANGES_REQUESTED** [review]

File: tools/dataset.cpp [hunk 5]
```diff
@@ -197,3 +266,62 @@ void ukv_graph_import(ukv_graph_import_t* c_ptr) {
...
+        ukv_graph_find_edges(&graph_find);
+
+        count = stream.keys_batch().size();
+        total_ids = std::transform_reduce(degrees, degrees + count, 0ul, plus, [](ukv_vertex_degree_t d) {
+            return d != ukv_vertex_degree_missing_k ? d : 0;
+        });
+        total_ids *= 3;
+
+        export_method(c, max_batch_size, ids_in_edges, total_ids);
+        stream.seek_to_next_batch().throw_unhandled();
```
Comment (ashvardanian) [inline tools/dataset.cpp:326]: These methods shouldn’t throw any exceptions

File: tools/dataset.cpp [hunk 6]
```diff
@@ -197,3 +266,62 @@ void ukv_graph_import(ukv_graph_import_t* c_ptr) {
...
+void ukv_graph_export(ukv_graph_export_t* c_ptr) {
+
+    ukv_graph_export_t& c = *c_ptr;
+
+    auto export_method = &export_parquet;
+
+    ////// Choosing a method will look like this //////
+    //
+    // auto ext = c.paths_extension;
+    // auto export_method = ext == ".parquet" //
```
Comment (ashvardanian) [inline tools/dataset.cpp:280]: This won’t work, you are comparing pointers

File: tools/dataset.cpp (hunk 4 above)
Comment (Arman-Ghazaryan) [reply tools/dataset.cpp:293]: For small data, this may be overkill.

File: tools/dataset.cpp (hunk 6 above)
Comment (Arman-Ghazaryan) [reply tools/dataset.cpp]: Fixed

File: tools/dataset.cpp (hunk 5 above)
Comment (Arman-Ghazaryan) [reply tools/dataset.cpp]: Fixed

File: tools/dataset.cpp (hunk 3 above)
Comment (Arman-Ghazaryan) [reply tools/dataset.cpp]: Fixed

File: tools/dataset.cpp (hunk 4 above)
Comment (ashvardanian) [reply tools/dataset.cpp]: These tools are only intended for large imports

File: src/modality_docs.cpp [hunk 7]
```diff
@@ -1281,6 +1287,30 @@ void ukv_docs_write(ukv_docs_write_t* c_ptr) {
     linked_memory_lock_t arena = linked_memory(c.arena, c.options, c.error);
     return_on_error(c.error);
 
+    std::vector<ukv_key_t> keys_vec;
```
Comment (ashvardanian) [inline src/modality_docs.cpp:1290]: No `std::vector`-s allowed... We have to use our own memory.

File: tools/dataset.cpp [hunk 8]
```diff
@@ -1 +1,1305 @@
...
+/////////  Helpers  /////////
+
+class arrow_visitor {
+  public:
+    arrow_visitor(std::string& json) : json(json) {}
+    arrow::Status Visit(arrow::NullArray const& arr) {
+        return arrow::Status(arrow::StatusCode::TypeError, "Not supported type");
+    }
+    arrow::Status Visit(arrow::BooleanArray const& arr) {
+        json = fmt::format("{}{},", json, arr.Value(idx));
```
Comment (ashvardanian) [inline tools/dataset.cpp:63]: This is horrible. 
You are reallocating a new string every time you want to append a boolean.
At least use `fmt::format_to`.

File: tools/dataset.cpp [hunk 9]
```diff
@@ -1 +1,1305 @@
...
+using graph_t = std::vector<edge_t>;
+using docs_t = std::vector<value_view_t>;
+using fields_t = strided_iterator_gt<ukv_str_view_t const>;
+
+using ids_t = std::vector<std::pair<ukv_key_t*, ukv_size_t>>;
+using vals_t = std::vector<std::pair<ukv_bytes_ptr_t, ukv_size_t>>;
+
+/////////  Helpers  /////////
+
+class arrow_visitor {
```
Comment (ashvardanian) [inline tools/dataset.cpp:56]: Check styling guidelines.

File: tools/dataset.cpp [hunk 10]
```diff
@@ -1 +1,1305 @@
...
+    }
+    arrow::Status Visit(arrow::DenseUnionArray const& arr) {
+        return arrow::Status(arrow::StatusCode::TypeError, "Not supported type");
+    }
+
+    std::string& json;
+    size_t idx = 0;
+};
+
+bool strcmp_(const char* lhs, const char* rhs) {
```
Comment (ashvardanian) [inline tools/dataset.cpp:226]: Style guidelines: East const

File: tools/dataset.cpp [hunk 11]
```diff
@@ -1 +1,1305 @@
...
+    std::shared_ptr<arrow::Array> keys_array;
+    std::shared_ptr<arrow::Array> docs_array;
+    std::vector<ukv_key_t> keys_vec;
+    std::vector<std::string> docs_vec;
+
+    keys_vec.reserve(size);
+    docs_vec.reserve(size);
+
+    if (c.fields) {
+        std::vector<std::string> fields(c.fields_count);
```
Comment (ashvardanian) [inline tools/dataset.cpp:1160]: Why do you need a `std::vector` or `std::string`?

File: tools/dataset.cpp [hunk 12]
```diff
@@ -1 +1,1305 @@
...
+void export_ndjson_d( //
+    ukv_docs_export_t& c,
+    std::vector<ptr_range_gt<ukv_key_t const>> const& keys,
+    fields_t const& strided_fields,
+    ukv_size_t size_in_bytes,
+    vals_t const& values) {
+
+    char file_name[uuid_length];
+    make_uuid(file_name);
+    std::ofstream output(fmt::format("{}{}", file_name, c.paths_extension));
```
Comment (ashvardanian) [inline tools/dataset.cpp:1207]: Please, never use `ofstream` or the rest of the old-school IO libraries from STL. Especially on the hot path.

File: tools/dataset.cpp (hunk 2 above)
Comment (Arman-Ghazaryan) [reply tools/dataset.cpp:443]: Its not working with parquet.

**Pull request closed without merging**